### PR TITLE
[INFINITY-2364] Move from suppress/revive to decline/revive logic

### PR DESCRIPTION
**Pull request by gabrielhartmann**
Instead of suppressing offers when all work is complete, we now set a decline timeout of 2 weeks (a.k.a. forever) whenever we suppress any offer.  When we see *new* work (`PodInstanceRequirement`s) we assume that the offers we've declined forever may be useful to that work, and so we revive offers.

Pseudo-code algorithm is like this:
```
// We always start out revived
List<Requirement> currRequirements = getRequirements();

while (true) {
    List<Requirement> newRequirements = getRequirements() - currRequirements;
    if (newRequirements.isEmpty()) {
        print “No new work”;
    } else {
        currRequirements = newRequirements;
        revive();
    }
}

```

I've left out the `supress` portion of the logic in the pseudocode, but it's in the real code.  Basically if there's no work, suppress.

A natural question is why do we overwrite `currRequirements` with `newRequirements`?  Here's the invariant.  Anything that is in `currRequirements` has had revive called for it.  Any change with regard to `currRequirements` i.e. `newRequirements` implies that those requirements may need an offer which has been declined forever, so we need to revive.  We cannot maintain a cummulative list in `currRequirements` because we may see the same work twice.

e.g. 
```
1. `kafka-0-broker` fails    @ 10:30, it's new work!
2. `kafka-0-broker` recovers @ 10:35
3. `kafka-0-broker` fails    @ 11:00, it's new work!
...
```

### CODE REVIEW

**Comment by benclarkwood** [comment]
[Move from suppress to decline](https://trello.com/c/NaIAwk3x/9-move-from-suppress-to-decline)

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java [hunk 1]
```diff
@@ -234,15 +234,12 @@ protected void postRegister() {
...
-                    stateStore,
-                    configStore,
-                    driver,
-                    eventBus,
-                    getPlanManagers());
+            suppressReviveManager = new SuppressReviveManager(driver, stateStore, getPlanCoordinator());
         }
 
-        suppressReviveManager.start();
+        if (processingOffers.compareAndSet(false, true)) {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: May be worth a comment that this expects to be called by both register() and reregister(). Should also mention that we want to re-invoke reconcilation each time but that the offer processing loop doesn't need to be restarted.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/SuppressReviveManager.java [hunk 2]
```diff
@@ -92,161 +63,60 @@ public void run() {
...
     }
 
     private void revive() {
-        synchronized (suppressReviveLock) {
-            setOfferMode(false);
-        }
+        setOfferMode(false);
     }
 
     private void setOfferMode(boolean suppressed) {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/SuppressReviveManager.java]: `setSuppressed(bool)`?

**Comment by nickbp** [comment]
FYI `template` appears to be failing to uninstall in CI. Could be related to the PR.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/SuppressReviveManager.java [hunk 3]
```diff
@@ -92,161 +63,60 @@ public void run() {
...
-    }
+        logger.debug("Old candidates: {}", candidates);
+        logger.debug("New candidates: {}", newCandidates);
 
-    private void suppressOrRevive() {
-        boolean hasOperations = planManagers.stream()
-                .anyMatch(planManager -> PlanUtils.hasOperations(planManager.getPlan()));
-        if (hasOperations || hasTasksNeedingRecovery()) {
-            transitionState(State.WAITING_FOR_OFFER);
+        if (newCandidates.isEmpty()) {
```
Comment (benclarkwood) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/SuppressReviveManager.java]: IMO the comments attached to this PR as a description would make sense here as documentation.

**Comment by gabrielhartmann** [comment]
@nickbp:  The uninstall failures are due to offer suppression being aggressive.  We suppress offers right away if we don't have any candidates which require offers.  Uninstall has one Step which needs to be executed which doesn't need offers so sometimes it gets unlucky and doesn't get executed.  See #1594 for a solution.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/OfferUtils.java [hunk 4]
```diff
@@ -14,6 +14,10 @@
  */
 public class OfferUtils {
     private static final Logger LOGGER = LoggerFactory.getLogger(OfferUtils.class);
+    private static final int TWO_WEEKS_S = 2 * 7 * 24 * 60 * 60;
```
[the remark this answers is not in the document]
Comment (elezar) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/OfferUtils.java]: Or:
```java
private static final long TWO_WEEKS_S = Duration.ofDays(14).getSeconds();
```
(but then it's long).

[the remark this answers is not in the document]
Comment (nickbp) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/OfferUtils.java]: Duration version looks good IMO

[the remark this answers is not in the document]
Comment (benclarkwood) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/OfferUtils.java]: Yeah, I forgot about Duration. But remembered that TimeUnit existed.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/SuppressReviveManager.java [hunk 5]
```diff
@@ -92,161 +63,60 @@ public void run() {
...
                 pollInterval,
                 TimeUnit.SECONDS);
-        logger.info(
-                "Monitoring these plans for suppress/revive: {}",
-                planManagers.stream().map(planManager -> planManager.getPlan().getName()).collect(Collectors.toList()));
-    }
-
-    public void start() {
-        // This must be run on a separate thread to avoid deadlock with the MesosToSchedulerDriverAdapter
-        new Thread(new Runnable() {
```
[the remark this answers is not in the document]
Comment (gabrielhartmann) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/SuppressReviveManager.java]: yup

**Pull request merged**